validator/client/src/main/java/tech/pegasys/teku/validator/client/AbstractDutyScheduler.java

```diff
@@ -67,34 +63,37 @@ public void onChainReorg(final UInt64 newSlot, final UInt64 commonAncestorSlot)
     LOG.debug(
         "Chain reorganisation detected. Invalidating validator duties after epoch {}",
         lastUnaffectedEpoch);
-    removeEpochs(dutiesByEpoch.tailMap(lastUnaffectedEpoch, false));
-    recalculateDuties(compute_epoch_at_slot(newSlot));
+    invalidateEpochs(dutiesByEpoch.tailMap(lastUnaffectedEpoch, false));
+    calculateDuties(compute_epoch_at_slot(newSlot));
   }
 
   @Override
   public void onPossibleMissedEvents() {
     // We may have missed a re-org notification so we need to recalculate all duties.
-    removeEpochs(dutiesByEpoch);
-    currentEpoch.ifPresent(this::recalculateDuties);
+    invalidateEpochs(dutiesByEpoch);
   }
 
-  protected void recalculateDuties(final UInt64 epochNumber) {
-    dutiesByEpoch.computeIfAbsent(epochNumber, this::requestDutiesForEpoch);
+  private void calculateDuties(final UInt64 epochNumber) {
+    dutiesByEpoch.computeIfAbsent(epochNumber, this::createEpochDuties);
     for (int i = 1; i <= lookAheadEpochs; i++) {
-      dutiesByEpoch.computeIfAbsent(epochNumber.plus(i), this::requestDutiesForEpoch);
+      dutiesByEpoch.computeIfAbsent(epochNumber.plus(i), this::createEpochDuties);
     }
   }
 
-  protected void removePriorEpochs(final UInt64 epochNumber) {
-    final NavigableMap<UInt64, DutyQueue> toRemove = dutiesByEpoch.headMap(epochNumber, false);
-    removeEpochs(toRemove);
+  private EpochDuties createEpochDuties(final UInt64 epochNumber) {
+    return EpochDuties.calculateDuties(epochDutiesScheduler, epochNumber);
   }
 
-  protected void removeEpochs(final NavigableMap<UInt64, DutyQueue> toRemove) {
-    toRemove.values().forEach(DutyQueue::cancel);
+  private void removePriorEpochs(final UInt64 epochNumber) {
+    final NavigableMap<UInt64, EpochDuties> toRemove = dutiesByEpoch.headMap(epochNumber, false);
+    toRemove.values().forEach(EpochDuties::cancel);
```

Comment:
Why is there a need for canceling epoch duties of older epochs? Is this a defensive measure?

Reply:
If they're still waiting for the duties to load and have pending actions queued up we don't want to waste time executing them as they're now too old.